### PR TITLE
KOGITO-7484: Knative CLI Workflow Plugin II - deploy with operator

### PR DESCRIPTION
**Pull request by ederign**
## Goal
Expand Knative developer experience, allowing users to create, run and deploy single file workflows, hiding all Java and Quarkus complexity.

This PR will NOT build over our CI. We still depend on a release on [kogito-serverless-operator](https://github.com/kiegroup/kogito-serverless-operator). (check go.mod file my replaces)

This PR is for testing and review only:

## On this PR

This PR introduces the deploy command for single-file.

### How to test it ?
[Guide](https://docs.google.com/document/d/158FTFuCsI4vTGbgqMZjFaZbmVov302WZoNzVswcPCcQ/edit)

### What is pending, and will it come on the next PRs?

- [TP1 Blocker] Use kogito-serverless-operator fixed version
- [TP1 Blocker] = Use a fixed kogito dev mode image: "quay.io/kiegroup/kogito-swf-devmode-nightly:latest"  
- [TP1] Clean up jiras
- [TP1] Test integration with kncli -> I'll focus on this next
- [TP1] Run more complex/complet single file projects
- [TP1] Upgrade docs (it will be done after freeze of TP1)

### CODE REVIEW

File: packages/kn-plugin-workflow/.gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 debug/
-TODO.txt
\ No newline at end of file
+kogito-serverless-operator/
+TODO.txt
```
Comment (caponetto) [inline packages/kn-plugin-workflow/.gitignore:3]: Is this file necessary?

File: packages/kn-plugin-workflow/go.mod [hunk 2]
```diff
@@ -2,13 +2,20 @@ module github.com/kiegroup/kie-tools/packages/kn-plugin-workflow
 
 go 1.19
 
+replace (
+	github.com/kiegroup/kogito-serverless-operator/api v0.0.0 => /Users/ederign/go/src/github.com/kiegroup/kogito-serverless-operator/api
```
Comment (caponetto) [inline packages/kn-plugin-workflow/go.mod]: This is not final, right? `/Users/ederign`

File: packages/kn-plugin-workflow/pkg/command/deploy.go [hunk 3]
```diff
@@ -20,45 +20,324 @@ import (
...
+	Delete                     bool
+	DeployingDeletingCapsLabel string
+	DeployedDeletingLabel      string
 }
 
 func NewDeployCommand() *cobra.Command {
 	var cmd = &cobra.Command{
 		Use:   "deploy",
-		Short: "Deploy a Kogito Serverless Workflow project",
+		Short: "Deploy a Kogito Serverless Workflow file on Kubernetes via Kogito Operator",
```
Comment (caponetto) [inline packages/kn-plugin-workflow/pkg/command/deploy.go]:
```suggestion
		Short: "Deploy a Kogito Serverless Workflow file on Kubernetes via Kogito Serverless Workflow Operator",
```

File: packages/kn-plugin-workflow/pkg/metadata/constants.go [hunk 4]
```diff
@@ -36,4 +37,13 @@ const (
 	WorkflowSwJson = "workflow.sw.json"
 
 	KogitoImage = "quay.io/kiegroup/kogito-swf-devmode-nightly:latest"
+
+	OperatorName       = "kogito-serverless-operator-system"
+	OperatorManagerPod = "kogito-serverless-operator-controller-manager"
+
+	YAMLExtension         = "sw.yaml"
```
Comment (caponetto) [inline packages/kn-plugin-workflow/pkg/metadata/constants.go:44]: Do we need to account for `sw.yml` too?

File: packages/kn-plugin-workflow/pkg/command/deploy.go [hunk 5]
```diff
@@ -20,45 +20,324 @@ import (
...
 		Long: `
-	Deploys a Kogito Serverless Workflow project in the current directory
-	in the Kogito operator. 
+	Deploy a Kogito Serverless Workflow file in Kubernetes via the Kogito Operator.
```
Comment (caponetto) [inline packages/kn-plugin-workflow/pkg/command/deploy.go]:
```suggestion
	Deploy a Kogito Serverless Workflow file in Kubernetes via the Kogito Serverless Workflow Operator. 
```

File: packages/kn-plugin-workflow/pkg/command/deploy.go [hunk 6]
```diff
@@ -20,45 +20,324 @@ import (
...
+				fmt.Errorf("❌  failed to remove temp dir: %v", err)
+			}
+		}
+	}(&cfg)
+
+	if err != nil {
+		return fmt.Errorf("❌ initializing deploy config: %w", err)
+	}
+
+	fmt.Printf("🛠️️  %s a Kogito Serverless Workflow file on Kubernetes via the Kogito Operator...\n", cfg.DeployingDeletingCapsLabel)
```
Comment (caponetto) [inline packages/kn-plugin-workflow/pkg/command/deploy.go]:
```suggestion
	fmt.Printf("🛠️️  %s a Kogito Serverless Workflow file on Kubernetes via the Kogito Serverless Workflow Operator...\n", cfg.DeployingDeletingCapsLabel)
```

File: packages/kn-plugin-workflow/pkg/command/quarkus/run.go [hunk 7]
```diff
@@ -99,7 +99,7 @@ func runQuarkusProjectDevMode(cfg RunCmdConfig) (err error) {
 	go func() {
 		defer wg.Done()
 		if err := common.RunCommand(create, "mvn quarkus:dev"); err != nil {
-			fmt.Printf("❌ Error running Quarkus project: %v", err)
+			fmt.Printf("❌ ERROR:  running Quarkus project: %v", err)
```
Comment (caponetto) [inline packages/kn-plugin-workflow/pkg/command/quarkus/run.go]:
```suggestion
			fmt.Printf("❌ ERROR: running Quarkus project: %v", err)
```

File: packages/kn-plugin-workflow/pkg/command/quarkus/create.go [hunk 8]
```diff
@@ -84,16 +84,16 @@ func runCreate() error {
 		return err
 	}
 
-	fmt.Println("🔨 Creating a Quarkus Kogito Serverless Workflow project...")
+	fmt.Println("🛠️ Creating a Quarkus Kogito Serverless Workflow project...")
 	if err = CreateQuarkusProject(cfg); err != nil {
-		fmt.Println("❌  Error creating Quarkus project", err)
+		fmt.Println("❌ERROR: creating Quarkus project", err)
```
Comment (caponetto) [inline packages/kn-plugin-workflow/pkg/command/quarkus/create.go]:
```suggestion
		fmt.Println("❌ ERROR: creating Quarkus project", err)
```

File: packages/kn-plugin-workflow/pkg/command/deploy.go [hunk 9]
```diff
@@ -20,45 +20,324 @@ import (
...
-	"Meanwhile, you can deploy your project via \"quarkus build\" and \"quarkus deploy\" commands.\n" +
-	"To convert it to Quarkus, run \"quarkus convert\" command.\n"
+func runDeployUndeploy(cmd *cobra.Command, args []string) error {
+
+	cfg, err := runDeployCmdConfig(cmd)
+	//temp dir cleanup
+	defer func(cfg *DeployCmdConfig) {
+		if cfg.TempDir != "" {
+			if err := os.RemoveAll(cfg.TempDir); err != nil {
+				fmt.Errorf("❌  failed to remove temp dir: %v", err)
```
Comment (caponetto) [inline packages/kn-plugin-workflow/pkg/command/deploy.go]: Should we include the `❌ ERROR:` prefix in all error messages like in the other files?

File: packages/kn-plugin-workflow/pkg/common/io.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+)
+
+func FindFilesWithExtensions(directoryPath string, extensions []string) ([]string, error) {
+	filePaths := []string{}
+
+	_, err := os.Stat(directoryPath)
+	if os.IsNotExist(err) {
+		return filePaths, nil
+	} else if err != nil {
+		return nil, fmt.Errorf("failed to access directory: %s", err)
```
Comment (caponetto) [inline packages/kn-plugin-workflow/pkg/common/io.go]: Should we include the `❌ ERROR:` prefix in all error messages like in the other files?

**Review by caponetto — COMMENTED** [review]
Very minor comments inline. I will give it a try shortly.

File: packages/kn-plugin-workflow/.gitignore (hunk 1 above)
Comment (ederign) [reply packages/kn-plugin-workflow/.gitignore:3]: It isn't, but doesn't hurt!  :)

File: packages/kn-plugin-workflow/go.mod (hunk 2 above)
Comment (ederign) [reply packages/kn-plugin-workflow/go.mod]: Operator guys just created a tag for me. updated!

File: packages/kn-plugin-workflow/pkg/metadata/constants.go (hunk 4 above)
Comment (ederign) [reply packages/kn-plugin-workflow/pkg/metadata/constants.go:44]: nice catch. I'll workon that.

File: packages/kn-plugin-workflow/pkg/common/io.go (hunk 10 above)
Comment (ederign) [reply packages/kn-plugin-workflow/pkg/common/io.go]: I'll fix it!

File: packages/kn-plugin-workflow/pkg/command/deploy.go (hunk 9 above)
Comment (ederign) [reply packages/kn-plugin-workflow/pkg/command/deploy.go]: I'll fix it!

Comment (ederign) [reply packages/kn-plugin-workflow/pkg/command/deploy.go]: fixed!

File: packages/kn-plugin-workflow/pkg/common/io.go (hunk 10 above)
Comment (ederign) [reply packages/kn-plugin-workflow/pkg/common/io.go]: fixed!

File: packages/kn-plugin-workflow/.gitignore (hunk 1 above)
Comment (caponetto) [reply packages/kn-plugin-workflow/.gitignore:3]: 🙈

**Review by caponetto — APPROVED** [review]
I have tested and looks good to me.
My only observation is that the building operation is too slow on my mac m1 but this should be fixed once we get a dev mode image for arm.

**Review by paulovmr — APPROVED** [review]

File: packages/kn-plugin-workflow/pkg/command/deploy.go [hunk 11]
```diff
@@ -20,45 +20,324 @@ import (
...
-	"Meanwhile, you can deploy your project via \"quarkus build\" and \"quarkus deploy\" commands.\n" +
-	"To convert it to Quarkus, run \"quarkus convert\" command.\n"
+func runDeployUndeploy(cmd *cobra.Command, args []string) error {
+
+	cfg, err := runDeployCmdConfig(cmd)
+	//temp dir cleanup
+	defer func(cfg *DeployCmdConfig) {
+		if cfg.TempDir != "" {
+			if err := os.RemoveAll(cfg.TempDir); err != nil {
+				fmt.Errorf("❌ ERROR:  failed to remove temp dir: %v", err)
```
Comment (zdrapela) [inline packages/kn-plugin-workflow/pkg/command/deploy.go]:
```suggestion
				fmt.Errorf("❌ ERROR: failed to remove temp dir: %v", err)
```

File: packages/kn-plugin-workflow/pkg/command/deploy.go [hunk 12]
```diff
@@ -20,45 +20,324 @@ import (
...
+			}
+		}
+	}(&cfg)
+
+	if err != nil {
+		return fmt.Errorf("❌ ERROR: initializing deploy config: %w", err)
+	}
+
+	fmt.Printf("🛠️️  %s a Kogito Serverless Workflow file on Kubernetes via the Kogito Serverless Workflow Operator...\n", cfg.DeployingDeletingCapsLabel)
```
Comment (zdrapela) [inline packages/kn-plugin-workflow/pkg/command/deploy.go]:
```suggestion
	fmt.Printf("🛠️️ %s a Kogito Serverless Workflow file on Kubernetes via the Kogito Serverless Workflow Operator...\n", cfg.DeployingDeletingCapsLabel)
```

File: packages/kn-plugin-workflow/pkg/command/deploy.go [hunk 13]
```diff
@@ -20,45 +20,324 @@ import (
...
+	//setup namespace
+	if len(cfg.NameSpace) == 0 {
+		if defaultNamespace, err := common.GetKubectlNamespace(); err == nil {
+			cfg.NameSpace = defaultNamespace
+		} else {
+			return err
+		}
+	}
+
+	fmt.Println("🔎  Checking if the Kogito Serverless Workflow Operator is correctly installed...")
```
Comment (zdrapela) [inline packages/kn-plugin-workflow/pkg/command/deploy.go]:
```suggestion
	fmt.Println("🔎 Checking if the Kogito Serverless Workflow Operator is correctly installed...")
```

File: packages/kn-plugin-workflow/pkg/command/deploy.go [hunk 14]
```diff
@@ -20,45 +20,324 @@ import (
...
+		return fmt.Errorf("❌ ERROR: failed to get current directory: %w", err)
+	}
+
+	applicationPropertiesPath := findApplicationPropertiesPath(dir)
+	if applicationPropertiesPath != "" {
+		cfg.ApplicationPropertiesPath = applicationPropertiesPath
+		fmt.Printf(" - ✅ Properties file found: %s\n", cfg.ApplicationPropertiesPath)
+	}
+
+	extensions := []string{".json", ".yaml"}
```
Comment (zdrapela) [inline packages/kn-plugin-workflow/pkg/command/deploy.go]: Should we also account for `.yml`?
```suggestion
	extensions := []string{".json", ".yaml", ".yml"}
```

File: packages/kn-plugin-workflow/pkg/command/deploy.go [hunk 15]
```diff
@@ -20,45 +20,324 @@ import (
...
+		fmt.Printf(" - ✅ Manifest %s successfully %s in namespace %s\n", path.Base(file), cfg.DeployedDeletingLabel, cfg.NameSpace)
+
+	}
+	return nil
+}
+
+func deploy(cfg *DeployCmdConfig) error {
+	fmt.Printf("🛠 Deploying your Kogito Serverless project in namespace %s\n", cfg.NameSpace)
+
+	manifestExtension := []string{".yaml"}
```
Comment (zdrapela) [inline packages/kn-plugin-workflow/pkg/command/deploy.go:145]: Should we also account for `.yml`?
```suggestion
	manifestExtension := []string{".yaml", ".yml"}
```

File: packages/kn-plugin-workflow/pkg/common/operator.go [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+	cmd := exec.Command("kubectl",
+		"delete",
+		"-f", crd,
+		"-n", namespace)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	err := cmd.Run()
+	if err != nil {
+		return fmt.Errorf("❌ ERROR: failed to execute kubectl apply command for %s: %s", crd, err)
```
Comment (zdrapela) [inline packages/kn-plugin-workflow/pkg/common/operator.go]:
```suggestion
		return fmt.Errorf("❌ ERROR: failed to execute kubectl delete command for %s: %s", crd, err)
```

File: packages/kn-plugin-workflow/pkg/common/operator.go [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+			return fmt.Errorf("❌ ERROR: failure accessing a path %q: %v\n", path, err)
+		}
+
+		if info.IsDir() || filepath.Ext(path) != ".yaml" {
+			return nil
+		}
+
+		file, err := os.Open(path)
+		if err != nil {
+			return fmt.Errorf("❌ ERROR:failure opening file %q: %v\n", path, err)
```
Comment (zdrapela) [inline packages/kn-plugin-workflow/pkg/common/operator.go]:
```suggestion
			return fmt.Errorf("❌ ERROR: failure opening file %q: %v\n", path, err)
```

File: packages/kn-plugin-workflow/pkg/common/operator.go [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+		}
+		defer file.Close()
+
+		byteValue, err := io.ReadAll(file)
+		if err != nil {
+			return fmt.Errorf("❌ ERROR:failure reading file %q: %v\n", path, err)
```
Comment (zdrapela) [inline packages/kn-plugin-workflow/pkg/common/operator.go]:
```suggestion
			return fmt.Errorf("❌ ERROR: failure reading file %q: %v\n", path, err)
```

**Review by zdrapela — APPROVED** [review]
Please see my inline comments. Those are just minor typos and consider accounting for manifest with `.yml`. (all suggested)

Once fixed, we are good to go. Looking good 👍️

File: packages/kn-plugin-workflow/pkg/command/deploy.go (hunk 15 above)
Comment (ederign) [reply packages/kn-plugin-workflow/pkg/command/deploy.go:145]: Manifests files are generate by operators api, so they are only YAML.

**Comment by ederign** [comment]
@zdrapela @caponetto thank you so much for the review!
